src/qibo/core/terms.py

```diff
@@ -0,0 +1,273 @@
+
+    Args:
+        matrix (np.ndarray): Full matrix corresponding to the term representation
+            in the computational basis. Has size (2^n, 2^n) where n is the
+            number of target qubits of this term.
+        q (int): Target qubit ids.
+    """
+
+    def __init__(self, matrix, *q):
+        self.target_qubits = tuple(q)
+        self._gate = None
+        self.hamiltonian = None
+        if not (matrix is None or isinstance(matrix, K.qnp.numeric_types) or
+                isinstance(matrix, K.qnp.tensor_types)):
+            raise_error(TypeError, "Invalid type {} of symbol matrix."
+                                   "".format(type(matrix)))
+        self._matrix = matrix
+
+    @property
+    def matrix(self):
+        return self._matrix
+
+    @property
+    def gate(self):
+        """Qibo gate that implements the action of the term on states."""
+        if self._gate is None:
+            self._gate = gates.Unitary(self.matrix, *self.target_qubits)
+        return self._gate
+
+    def exp(self, dt):
+        """Matrix exponentiation of the term."""
+        return K.qnp.expm(-1j * dt * self.matrix)
+
+    def expgate(self, dt):
+        """Unitary gate implementing the matrix exponentiation of the term."""
+        return gates.Unitary(self.exp(dt), *self.target_qubits)
+
+    def merge(self, term):
+        """Creates a new term by merging the given term to the current one.
+
+        The target qubits of the given term should be a subset of the target
+        qubits of the current term.
+        """
+        matrix = K.np.kron(term.matrix, K.eye(2 ** (len(self) - len(term))))
+        matrix = K.np.reshape(matrix, 2 * len(self) * (2,))
+        order = []
+        i = len(term)
+        for qubit in self.target_qubits:
+            if qubit in term.target_qubits:
+                order.append(term.target_qubits.index(qubit))
+            else:
+                order.append(i)
+                i += 1
+        order.extend([x + len(order) for x in order])
+        matrix = K.np.transpose(matrix, order)
+        matrix = K.np.reshape(matrix, 2 * (2 ** len(self),))
+        return HamiltonianTerm(self.matrix + matrix, *self.target_qubits)
+
+    def __len__(self):
+        return len(self.target_qubits)
+
+    def __mul__(self, x):
+        return HamiltonianTerm(x * self.matrix, *self.target_qubits)
+
+    def __rmul__(self, x):
+        return self.__mul__(x)
+
+    def __call__(self, state, density_matrix=False):
+        """Applies the term on a given state vector or density matrix."""
+        if density_matrix:
+            self.gate.density_matrix = True
+            return self.gate.density_matrix_half_call(state)
+        return self.gate(state) # pylint: disable=E1102
+
+
+class SymbolicTerm(HamiltonianTerm):
+    """:class:`qibo.core.terms.HamiltonianTerm` constructed from ``sympy`` symbols.
+
+    Example:
+        ::
+
+            from qibo.symbols import X, Y, SymbolicTerm
```

Comment:
I think the `SymbolicTerm` is not exposed by `qibo.symbols`.

Reply:
That is correct, this is a leftover from a previous version. The `SymbolicTerm` is now only accessed via `qibo.core.terms` and is not exposed to the user.